libtransmission/api-compat.cc

```diff
@@ -0,0 +1,843 @@
+    auto const was_legacy_response = was_legacy && src_top->contains(TR_KEY_result);
+    auto const is_response = was_jsonrpc_response || was_legacy_response;
+    auto const is_rpc = is_request || is_response;
+
+    auto state = detail::CloneState{};
+    state.style = tgt_style;
+    state.is_rpc = is_request || is_response;
+
+    auto const is_success = is_response &&
+        (was_jsonrpc_response ? src_top->contains(TR_KEY_result) :
+                                src_top->value_if<std::string_view>(TR_KEY_result).value_or("") == "success");
+
+    if (auto const method = src_top->value_if<std::string_view>(TR_KEY_method))
+    {
+        auto const key = tr_quark_convert(tr_quark_new(*method));
+        state.is_torrent = key == TR_KEY_torrent_get || key == TR_KEY_torrent_set;
+    }
+
+    if (is_response)
+    {
+        if (auto const* const args = src_top->find_if<tr_variant::Map>(was_jsonrpc ? TR_KEY_result : TR_KEY_arguments))
+        {
+            state.is_free_space_response = args->contains(TR_KEY_path) &&
+                args->contains(was_jsonrpc ? TR_KEY_size_bytes : TR_KEY_size_bytes_kebab);
+        }
+    }
+
+    auto ret = convert_impl(src, state);
+
+    auto* const tgt_top = ret.get_if<tr_variant::Map>();
+
+    // jsonrpc <-> legacy rpc conversion
+    if (is_rpc)
+    {
+        auto const is_jsonrpc = tgt_style == Style::Tr5;
+        auto const is_legacy = tgt_style != Style::Tr5;
+
+        // - use `jsonrpc` in jsonrpc, but not in legacy
+        // - use `id` in jsonrpc; use `tag` in legacy
+        if (is_jsonrpc)
+        {
+            tgt_top->try_emplace(TR_KEY_jsonrpc, tr_variant::unmanaged_string(JsonRpc::Version));
+            tgt_top->replace_key(TR_KEY_tag, TR_KEY_id);
+        }
+        else
+        {
+            tgt_top->erase(TR_KEY_jsonrpc);
+            tgt_top->replace_key(TR_KEY_id, TR_KEY_tag);
+        }
+
+        if (is_response && is_legacy && is_success && was_jsonrpc)
+        {
+            // in legacy messages:
+            // - move `result` to `arguments`
+            // - add `result: "success"`
+            tgt_top->replace_key(TR_KEY_result, TR_KEY_arguments);
+            tgt_top->try_emplace(TR_KEY_result, tr_variant::unmanaged_string("success"));
+        }
+
+        if (is_response && is_legacy && !is_success)
+        {
+            // in legacy error responses:
+            // - copy `error.data.error_string` to `result`
+            // - remove `error` object
+            // - add an empty `arguments` object
+            if (auto const* error = tgt_top->find_if<tr_variant::Map>(TR_KEY_error))
+            {
+                if (auto const* data = error->find_if<tr_variant::Map>(TR_KEY_data))
+                {
+                    if (auto const* errmsg = data->find_if<std::string_view>(TR_KEY_error_string_camel))
+                    {
+                        tgt_top->try_emplace(TR_KEY_result, *errmsg);
```

Comment:
If `error.data.error_string` is not available, you should fall back to `error.message`.

Example:

https://github.com/transmission/transmission/blob/64268a3a7f14f18db43184812dad91462e8604d5/web/src/rename-dialog.js#L90-L91